File: TwoSum/twoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        #assuming step one is  creating the hashmap
        prevMap = {}  # val -> index

        #iterate through nums array, record difference
        #if in hashmap, return indexes
        #otherwise, map a new value into the hashmap
        for i, n in enumerate(nums):
            diff = target - n
            if diff in prevMap:
                return [prevMap[diff], i]
            prevMap[n] = i

        """
        notes from video solution:
        type of problem: arrays and hashing
            first instinct is obviously to brute force with a nested loop, O(n^2)

        """

File: TwoSum/twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        //map each number to its index
        //if index - targen exists in hashmap, return indices
        //otherwise return nothing
        
        //making hashmap
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        int[] result = new int[2];
        
        for(int i = 0; i < nums.length; i++){  
            if(map.containsKey(target - nums[i])){
                result[0] = i;
                result[1] = map.get(target - nums[i]);
                break;
            } else {
                map.put(nums[i], i);
            }
            
        }
        
        return result;
    }
}
